utcp.c

```diff
@@ -577,41 +643,66 @@ static void ack(struct utcp_connection *c, bool sendatleastone) {
     pkt->hdr.ctl = ACK;
     pkt->hdr.aux = 0;
 
+    int err = 0;
     do {
         uint32_t seglen = left > c->utcp->mtu ? c->utcp->mtu : left;
         uint32_t bufpos = seqdiff(c->snd.nxt, c->snd.una);
         pkt->hdr.seq = c->snd.nxt;
 
-        c->snd.nxt += seglen;
         left -= seglen;
 
         // when FIN is not ack'ed yet len must be at least 1
-        if(seglen && fin_wanted(c, c->snd.nxt)) {
+        if(seglen && fin_wanted(c, c->snd.nxt + seglen)) {
             seglen--;
             pkt->hdr.ctl |= FIN;
         }
 
         buffer_copy(&c->sndbuf, pkt->data, bufpos, seglen);
 
+        size_t pktlen = sizeof pkt->hdr + seglen;
+        print_packet(c->utcp, "send", pkt, pktlen);
+        ssize_t sent = c->utcp->send(c->utcp, pkt, pktlen);
+        if(sent != pktlen) {
+            if(sent > pktlen) {
+                utcp_send_error(pkt, pktlen, sent, false);
+            }
+            else if(sent >= 0 || sent == UTCP_WOULDBLOCK) {
+                // when no data could be sent with possibly the header broken
+                // or when the socket would block, don't advance but retry later
+                utcp_send_error(pkt, pktlen, sent, false);
+                err = UTCP_WOULDBLOCK;
+                break;
```

Comment:
These `break` statements skip the `c->snd.next` updates below. Is that correct behaviour?